### ISSUE TITLE
`Connection lost exc=None` when sending message to empty group

### ISSUE THREAD

**Issue by marcel-robitail** [CJWorkbench/channels_rabbitmq#4]
When I try to send a message to an empty group with `get_channel_layer().group_send()`, I get the following message:
```
Connection lost exc=None
```

I don't think it's a serious issue, but when I send a lot of messages it does spam my logs.

The message is logged in `aioamqp/protocol.py` at line 112 in `connection_lost()`. I know that this is not your package, but I can't reproduce this in `channels_redis`, leading me to believe that it can be fixed here.

I added a `traceback.print_stack()` in that function to make it easier to trace down the issue. The sensitive data has been removed.

```
    channel_layer.group_send(...)
  File ".../venv/lib/python3.6/site-packages/asgiref/sync.py", line 52, in __call__
    loop.close()
  File ".../venv/lib/python3.6/site-packages/channels_rabbitmq/core.py", line 100, in _wrapper
    self.run_until_complete(connection.close())
  File "/usr/lib/python3.6/asyncio/base_events.py", line 460, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.6/asyncio/base_events.py", line 427, in run_forever
    self._run_once()
  File "/usr/lib/python3.6/asyncio/base_events.py", line 1440, in _run_once
    handle._run()
  File "/usr/lib/python3.6/asyncio/events.py", line 145, in _run
    self._callback(*self._args)
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 662, in _call_connection_lost
    self._protocol.connection_lost(exc)
  File ".../venv/lib/python3.6/site-packages/aioamqp/protocol.py", line 114, in connection_lost
    traceback.print_stack()
```

I am using `channels==2.0.2` , `channels-rabbitmq==0.0.11`, and `asgiref==2.3.2`.

**Comment by marcel-robitail** [comment]
As a workaround, I have set the log level of `aioamqp.protocol` to `CRITICAL`.

**Comment by adamhooper** [comment]
Thanks for the bug report! I can't manage to reproduce this -- maybe because we're using different code.

I tried adding this unit-test, and it passed:

```
@ASYNC_TEST
async def test_groups_no_such_group(connect):
    """
    Tests that group_send does nothing if there is no such group
    """
    connection = connect("x")
    # This group doesn't exist: the message should disappear
    await connection.group_send("my-group", {"type": "message.1"})

    # Now create the group, and check that new messages are delivered
    await connection.group_add("my-group", "x!1")
    await connection.group_send("my-group", {"type": "message.2"})
    assert (await connection.receive("x!1"))["type"] == "message.2"
```

I'll need more information. I guess the first step is: could you please share the lines of code around your invocation of group_send()? (At a minimum, it should answer: What's the method signature? Are you using await? loop.create_task()? What awaits the task?)

**Comment by marcel-robitail** [comment]
Sorry, I should have specified. It's the `.save()` method on a Django model that call `group_send` (so it's synchronous). It looks like this:
```
def save(self, *args, **kwargs):
    super(ClassName, self).save(*args, **kwargs)

    channel_layer = channels.layers.get_channel_layer()
    asgiref.sync.async_to_sync(channel_layer.group_send)(
        'group-that-doesnt-exist',
        {'type': 'message.1'},
    )
```

I will run your test case and see if I can tweak it to reproduce the issue.

**Comment by adamhooper** [comment]
First, a hunch: I don't know why this only happens when a group doesn't exist. As far as I know, the bytes travelling across wires should be identical. Maybe RabbitMQ is 1ms faster when the group doesn't exist, and that changes timings on the Python side?

Second, some context on `async_to_sync()`: it's evil. It spins up a new event loop on every call. With a new event loop comes a new RabbitMQ connection; so every time you call async_to_sync(channel_layer.group_send)(...), that means: 1. create new event loop; 2. connect to RabbitMQ; 3. await group_send(); 4. shut down the event loop (which is a whole set of hacks unto itself).

I just noticed your traceback shows the error during `connection.close()`. That means the error comes during event-loop shutdown, not during send. On the bright side, your workaround ("ignore the exception") is probably fine.

But we need to know how to reproduce this. We on production are also using `async_to_sync()`, and we've never experienced this problem.

The best way to debug is to "reduce" your web app. This can be fun; it might take about 20 minutes:

1. Start with a fresh checkout, and build a set of instructions that produces the exception 100% of the time.
2. Delete _all_ the files and imports you can. (For instance, delete the code to every endpoint that isn't in step 1.) Test that you can still cause the error using the instructions from step 1.
3. Delete all the dependencies you no longer need (because you deleted so many files). Test that you can still cause the error using the instructions from step 1.
4. If you're left with more than one file, return to step 2.
5. Attach your reduction here, or on a GitHub gist.

Oh, and: what version of Python are you using? Do you get the same behavior with Python 3.7.3?

**Comment by marcel-robitail** [comment]
Thanks for the reply. As a primarily node.js developer who is used to using promises for *everything*, `async_to_sync()` also seems very wrong to me. My consumers are async, but I did a quick search and I didn't find a way to make a Django model's `save()` method async. Do you have any suggestions on how to avoid `async_to_sync()` in this case? Maybe handle it through celery?

I am working on the "reduced" version.

**Comment by marcel-robitail** [comment]
Another thing I missed, it's only when I call `async_to_sync()` in a thread.

Here is the minimal script that gives mes the error:
```python
import os
import sys
from threading import Thread
from django.apps import AppConfig
from channels.layers import get_channel_layer
from asgiref.sync import async_to_sync


class TestApp(AppConfig):
    def _run_test(self):

        channel_layer = get_channel_layer()

        async_to_sync(channel_layer.group_send)('my-group', {
            'type': 'message.1',
            'message': 'Hello World!',
        })

    def ready(self):
        if 'runserver' not in sys.argv or os.getenv('RUN_MAIN') != 'true':
            return

        thread = Thread(
            target=self._run_test,
            daemon=True,
        )
        thread.start()

```

And the stack trace:
```
File "/usr/lib/python3.6/threading.py", line 884, in _bootstrap
    self._bootstrap_inner()
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File ".../apps.py", line 18, in _run_test
    'message': 'Hello World!',
  File ".../venv/lib/python3.6/site-packages/asgiref/sync.py", line 52, in __call__
    loop.close()
  File ".../venv/lib/python3.6/site-packages/channels_rabbitmq/core.py", line 100, in _wrapper
    self.run_until_complete(connection.close())
  File "/usr/lib/python3.6/asyncio/base_events.py", line 460, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.6/asyncio/base_events.py", line 427, in run_forever
    self._run_once()
  File "/usr/lib/python3.6/asyncio/base_events.py", line 1440, in _run_once
    handle._run()
  File "/usr/lib/python3.6/asyncio/events.py", line 145, in _run
    self._callback(*self._args)
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 662, in _call_connection_lost
    self._protocol.connection_lost(exc)
  File ".../venv/lib/python3.6/site-packages/aioamqp/protocol.py", line 114, in connection_lost
    traceback.print_stack()
```

I will try python 3.7 next.

**Comment by marcel-robitail** [comment]
Exact same behaviour on Python 3.7.3.

**Comment by adamhooper** [comment]
Ooh, we're getting somewhere. I'm glad you took the database out of the equation. Still more reduction to go: does this happen if your copy/pasted app is the _only_ Django app in your settings.py? If so, that sounds like something I should be able to reproduce.

Also, my django-channels memory is returning: I know how you can prevent those `loop.close()` calls -- which should make the error go away in your app. (I'd still love to solve the underlying issue, of course.)

If you're calling `model.save()` from an async method -- meaning somewhere, _anywhere_ in the stack there's an async function call -- you must wrap it in `@database_sync_to_async`. Yes, you _must_ wrap _every_ database read or write in `@database_sync_to_async()`, _everywhere_ it might be executed through an async function call -- otherwise, you'll create database connections on the wrong threads and they'll never be cleaned up, so eventually your DB server will run out of connections and/or it'll disconnect a connection that's been idle too long, causing an error Django isn't prepared for. (The most common one we've seen: database queries on a thread stop working until you restart the app manually.)

And then, if you're using `@database_sync_to_async()`, you don't need `async_to_sync()` at all.

Here are the main pieces -- these could be methods within a Django model, for instance....:

```
@database_sync_to_async
def _save_async(self, *args, **kwargs):
    """Save, asynchronously."""
    return self.save(self, *args, **kwargs):

async def save_and_notify(self, *args, **kwargs):
    await self._save_async()

    channel_layer = channels.layers.get_channel_layer()
    await channel_layer.group_send(
        'group-that-doesnt-exist',
        {'type': 'message.1'},
    )
```

And your handler must `await model.save_and_notify()` instead of calling `model.save()`.

Every. Single. Database read and/or write should look like this.

Where `async_to_sync()` spreads chaos, `sync_to_async()` (which `database_sync_to_async()` uses) reins it in a bit. `async_to_sync()` runs a sync function on the event loop's threadpool. _Plus_, it has the bonus feature of storing a reference to that event loop in thread-local storage -- meaning any `async_to_sync()` calls _nested within_ a `sync_to_async()` call actually use the original event loop instead of spawning another one.

In your example, you're definitely creating a new event loop when you call `async_to_sync()`. That's probably not your intent. And if you fix _that_, this bug will go away for you.

That would explain why you're seeing problems with new threads. Don't call `async_to_sync()` from within threads that aren't managed by `sync_to_async()`.

I commiserate: yes, this is way too complicated. No, I doubt it will get simpler. Heck, it gets even more complex because there's this whole thing about coroutine cancellation that you can't avoid. If I had to do it all over again, I'd use Node. (Our project is inextricably linked to Pandas. _Still_ I'd use Node.)

**Comment by marcel-robitail** [comment]
Yes, it is the only app in `settings.py`. The `urlpatterns` in the app's `urls.py` is also empty. I have deleted the app's `routing.py` and `consumers.py`.

Wow, you have taught me a lot about async operations in python. Thank you for the detailed explanation! Now if I have to `await model.save_and_notify()`, the function in my thread has to be async, right? How do I run my async function in a thread without shooting myself in the foot in the same way?

Unfortunately, Python was the only option for this project for reasons I wont get into. I love Python, but some parts still seem clunky.

Thanks again for all the information and sorry that my bug report has turned into you fixing my code.

**Comment by adamhooper** [comment]
@marcel-robitail I've created a Django app that uses your code. It didn't produce an error for me.

Could you (or anyone else experiencing this bug) please produce this error in a way I can replicate? For instance, a Docker image? I can only fix it once I replicate it.

**Comment by marcel-robitail** [comment]
I can reproduce it using this [minimal repository](https://github.com/marcel-robitail/channels_rabbitmq_issue4) on Ubuntu 18.10. I am not too familiar with Docker, but if you still can't reproduce it, I will try to create an image.

**Comment by adamhooper** [comment]
A-ha! I see it. Thanks.

**Comment by adamhooper** [comment]
Ah. It's https://github.com/Polyconseil/aioamqp/pull/155 ... fixed in `aioamqp` but awaiting a new release. I'll leave this issue open until the release happens.

The warning is, indeed, harmless.

That said, `django-channels` users in particular should never see this because _you shouldn't be disconnecting from RabbitMQ_ on production.

@marcel-robitail In your case, you're getting disconnected for two reasons:

1. `async_to_sync()` is called before Daphne's event loop is started. I surmise that makes `async_to_sync()` create a temporary event loop and then shut it down -- causing a disconnect. This happens in your reduction only -- not in your real-world code. (A `time.sleep(1)` within the thread circumvents this problem.)
2. Your thread isn't aware of the main event loop. If you're going to create custom threads that do async stuff, you need to pass them the main event loop. Then you can set `async_to_sync(...)`'s return value's `.main_event_loop` ... or better yet, skip `async_to_sync()` entirely and use `asyncio.run_coroutine_threadsafe()`. I've written out both ideas below.

```python
import sys
from threading import Thread
from django.apps import AppConfig
from channels.layers import get_channel_layer
from asgiref.sync import async_to_sync
from daphne.server import Server
from twisted.internet import reactor


class TheAppConfig(AppConfig):
    name = 'the_app'

    def _run_script(self, loop):
        channel_layer = get_channel_layer()

        # My preferred sending method:
        print('sending with asyncio')
        asyncio.run_coroutine_threadsafe(channel_layer.group_send('my-group', {
            'type': 'script.event',
            'message': 'Hello World!',
        }), loop).result()
        print('sent')

        # The async_to_sync() way
        import time
        time.sleep(1)  # Wait for Channels to start the event loop
        print('sending with async_to_sync()')
        sync_group_send = async_to_sync(channel_layer.group_send)
        sync_group_send.main_event_loop = loop
        sync_group_send('my-group', {
            'type': 'script.event',
            'message': 'Hello World!',
        })
        print('sent')

    def ready(self):
        if 'runserver' not in sys.argv or os.getenv('RUN_MAIN') != 'true':
            return

        asyncio.set_event_loop(reactor._asyncioEventloop)
        singleton_loop = asyncio.get_event_loop()
        thread = Thread(
            target=self._run_script,
            daemon=True,
            args=(singleton_loop,)
        )
        thread.start()
```

I've only been doing this async-in-Python stuff for a year. My advice: if you're going to start other threads, don't let them touch asyncio. Stick with the threads you can access through `asyncio.get_event_loop().run_in_executor(...)`, which is how asyncio is designed to interact with worker threads. If you're going to call `async_to_sync()`, call it within a `sync_to_async()` block.

Django views are called within `sync_to_async()`; so I think your use case is unique. All the same, when we upgrade aioamqp the error message will go away.

**Comment by marcel-robitail** [comment]
Thank you for the very detailed description. I really appreciate the time you've taken for this issue.

I think I understand the first part of what you said about don't use `async_to_sync` but rather `asyncio.run_coroutine_threadsafe`. I don't fully understand the second part about "don't let other threads touch asyncio". Should I never have a thread that calls `save_and_notify`? Did I completely misunderstand that?

**Comment by adamhooper** [comment]
@marcel-robitail `asyncio` is largely an _alternative_ to spawning your own threads. Ideally, if you're using asyncio you shouldn't _need_ to spawn your own threads. (node.js only added worker threads in version >10.)

In reality, the Python ecosystem is ... chaotic. Many big-name libraries (e.g., boto3, tweepy, django.db ...) are synchronous, so one typically wants to call them from within threads ... and await them. The simplest way is `EventLoop.run_in_executor(...)`, which delegates to a thread pool (as opposed to spawning your own threads manually). That's how `sync_to_async()` works.

Going the _other way around_ -- calling async code from within a sub-thread (even an executor thread) -- is error-prone.  `EventLoop.call_soon()` is not thread-safe -- one must use `EventLoop.call_soon_threadsafe()` or suffer _undefined behavior_. There's no guarantee of a singleton event loop. `async_to_sync()` and Daphne both paper over this flavor of problem and present simpler APIs, but some cases fall through the cracks (like your example code) and the result is confusion. It's not clear how to avoid this sort of problem.

All that to say: in my opinion, mixing threads and event loops is chaos.

(We do use it ourselves in [CJWorkbench](https://github.com/CJWorkbench/cjworkbench). I'm not saying "never do it." I'm saying: "it's hell." I suspect there's a nice way to avoid this sort of problem using queues ... but I'm not sure what that would look like.)

**Comment by marcel-robitail** [comment]
Unfortunately, not using threads for our use case isn't really an option. I am using `asyncio.run_coroutine_threadsafe` now. Everything is working and I am no longer getting those weird error messages. I really appreciate your spending the time to not only make my error disappear, but also to help me fix my misguided code.

I also have to save models outside my thread (in the thread setup method). I am using:
```python
asyncio.ensure_future(Model(
    foo='bar',
).save_and_notify(), loop=self.loop)
```
I don't know if this is the correct approach, but is seems to work.

**Comment by adamhooper** [comment]
Fixed in v1.1.0.
